### PR TITLE
[LLVMCPU] Enable CyclicNumProcsEqNumIters distribution for CPU backends.

### LINKED ISSUE #13205: Exceeded stack allocation limit on llvm-cpu
For

```mlir
hal.executable.variant public @embedded_elf_x86_64, target = <"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = false}> {
  hal.executable.export public @_main_dispatch_388_generic_1167x6x6x32x2x2_f32xf32xf32xi32 ordinal(0) layout(#hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer>]>]>) {
  ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index, %arg3: index, %arg4: index, %arg5: index, %arg6: index):
    %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg1, %arg2, %arg3, %arg4, %arg5, %arg6
    hal.return %x, %y, %z : index, index, index
  }
  builtin.module {
    func.func @_main_dispatch_388_generic_1167x6x6x32x2x2_f32xf32xf32xi32() {
      %c472775552 = arith.constant 472775552 : index
      %c499663232 = arith.constant 499663232 : index
      %c2 = arith.constant 2 : index
      %c-1_i32 = arith.constant -1 : i32
      %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c472775552) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<1167x12x12x32xf32>>
      %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c499663232) : !flow.dispatch.tensor<writeonly:tensor<1167x6x6x32xi32>>
      %2 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0, 0], sizes = [1167, 12, 12, 32], strides = [1, 1, 1, 1] : !flow.dispatch.tensor<readonly:tensor<1167x12x12x32xf32>> -> tensor<1167x12x12x32xf32>
      %3 = tensor.empty() : tensor<1167x6x6x32xi32>
      %4 = tensor.empty() : tensor<1167x6x6x32xf32>
      %5 = tensor.empty() : tensor<2x2xf32>
      %6 = linalg.fill ins(%c-1_i32 : i32) outs(%3 : tensor<1167x6x6x32xi32>) -> tensor<1167x6x6x32xi32>
      %7:2 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1 * 2 + d4, d2 * 2 + d5, d3)>, affine_map<(d0, d1, d2, d3, d4, d5) -> (d4, d5)>, affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel", "reduction", "reduction"]} ins(%2, %5 : tensor<1167x12x12x32xf32>, tensor<2x2xf32>) outs(%4, %6 : tensor<1167x6x6x32xf32>, tensor<1167x6x6x32xi32>) {
      ^bb0(%in: f32, %in_0: f32, %out: f32, %out_1: i32):
        %8 = arith.cmpf oge, %in, %out : f32
        %9 = arith.cmpi eq, %out_1, %c-1_i32 : i32
        %10 = arith.ori %8, %9 : i1
        %11 = linalg.index 4 : index
        %12 = linalg.index 5 : index
        %13 = arith.muli %11, %c2 : index
        %14 = arith.addi %13, %12 : index
        %15 = arith.index_cast %14 : index to i32
        %16 = arith.select %10, %15, %out_1 : i32
        %17 = arith.select %10, %in, %out : f32
        linalg.yield %17, %16 : f32, i32
      } -> (tensor<1167x6x6x32xf32>, tensor<1167x6x6x32xi32>)
      flow.dispatch.tensor.store %7#1, %1, offsets = [0, 0, 0, 0], sizes = [1167, 6, 6, 32], strides = [1, 1, 1, 1] : tensor<1167x6x6x32xi32> -> !flow.dispatch.tensor<writeonly:tensor<1167x6x6x32xi32>>
      return
    }
  }
}
```

I get

```
error: 'func.func' op exceeded stack allocation limit of 32768 bytes for function. Got 597504 bytes
```

### PR DESCRIPTION
**Pull request by hanhanW**
Distribution with more than three loops is supported. The commit sets other distributable loops to size one, which can avoid unbounded stack allocation.

Fixes https://github.com/openxla/iree/issues/13205

### CODE REVIEW

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by hanhanW** [comment]
hmm, looks like there would be regression on RISC-V side for MobileNet V1.

**Comment by hanhanW** [comment]
The dispatch sizes regression looks real. I looked into dumps and could not figure out why. The emulation looks good to me, it has different loop indices, which is expected. I sampled several dispatches and they all have same vector shapes. I haven't spotted suspicious dispatches that could regress the binary size, neither x86 nor risc-v.

The latency of MobileNetV1 on x86 is improved: 27.388 (vs. 29.525, 7.24%↓). I don't know the status about RISC-V. Do we want to avoid the regression about the dispatch sizes of MobileNetV1 on RISC-V 64? If so, @dcaballe can you help benchmark it or give me some pointers about debugging?

**Comment by benvanik** [comment]
(20% & 40% increases are big - would at least need to know why before landing - it'll only get harder to figure out)

**Comment by hanhanW** [comment]
One thing that I observed is `RemoveSingleIterationLoop` does not work if there are more than three distribution loops. The pass removes the distribution loops w/o the PR. This introduces more loops in the final IR, e.g., there could be four more loops (for distribution). Could it block some optimization on LLVM side, which increases binary sizes?

**Comment by dcaballe** [comment]
Yep, we need to figure this out. I would expect LLVM to remove a single iteration loop early on, unless the information to do so is no longer available at that level. I don't see unrolling having an impact here as the new loop is an outer one. Could you focus on one dispatch and share the assembly before and after? If you haven't spotted anything at MLIR, you may also want to share `--print-after-all` + flag to filter only one dispatch (https://reviews.llvm.org/D15776).

**Comment by hanhanW** [comment]
Okay, I'm pretty sure that the issue is because of distribution loops. I'm able to remove them entirely with `CyclicNumProcsEqNumIters`. There are no regression in terms of latency. We even improve total dispatch sizes. Let me see if I can fix the lit tests..

**Comment by hanhanW** [comment]
The PR triggers a bug in `CyclicNumProcsEqNumIters ` + `tile_interchange`. Interchange is not used in distribution AFAIK. It was added by me for experiments. The fix is not trivial to me, maybe we can drop the experimental test?

**Comment by hanhanW** [comment]
I changed additional logics after Mahesh's approval. I'll wait for at least one approval to land the PR.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:132]: I dont think we want to set this to 1 yet..... Can we leave it to the default 3....

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:132]: Leaving it to three has the same issue. The distribution loops are not removed and it regresses the total dispatch sizes. That's why I set it to one, which removes the loops when doing distribution.

image

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:132]: Should I remove those `RemoveSingleIterationLoop` passes with the PR as well?

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:132]: You should be able to set it to 3, and also use `linalg::DistributionMethod::CylciNumProcsEqNumIters`.

You can remove that pass from the CPU backends. I think VMVX still needs that pass (mostly because it is also relying on some loop canonicalizations that get brought in from that pass I think, we can look into that later)

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:132]: Setting it to 3 triggers other issues. We won't get static shapes, and they don't get vectorized. This regresses both execution time and total dispatch sizes.

IR dump before and after vectorization w/ setting to 3: https://gist.githubusercontent.com/hanhanW/3f1cb8797ae44c28b06189e657f210a4/raw

IR dump before and after vectorization  w/ setting to 1: https://gist.githubusercontent.com/hanhanW/eb93102ed10077bc5ea949be38828c9c/raw

**Comment by hanhanW** [comment]
The pad op test failed because `EliminateEmptyTensorsPass` can't handle the case. The tensor.empty op is not replaced with `flow.dispatch.tensor.load` in `eliminateEmptyTensors` function.

The IR before the patch: https://gist.githubusercontent.com/hanhanW/33c962af3a923be17768b3d7ac3f45d8/raw
The IR after the patch: https://gist.githubusercontent.com/hanhanW/992f3774031d937932cf0aa6273a78e6/raw

I took a deep look and found that this is because of domination issue. The offsets of the working version are all defined before the tensor.empty op, so we can replace it with a new op. The offsets of the later MLIR are defined right before `flow.dispatch.tensor.store` op, `mlir::bufferization::eliminateEmptyTensors` can't find an insertion point in this case. It is safe to move those operation to the insertion point. I'm not sure what the proper fix is. @matthias-springer can you help on this?

To repro, run `iree-opt --iree-eliminate-empty-tensors ~/a.mlir`

```mlir
    %23 = tensor.empty() : tensor<1x19x19x64xf32>
    %24 = affine.min affine_map<()[s0, s1] -> (0, s0 - s1)>()[%3, %2]
    %25 = tensor.empty() : tensor<1x1x1x4xf32>
    %26 = scf.for %arg0 = %c0 to %c19 step %c1 iter_args(%arg1 = %23) -> (tensor<1x19x19x64xf32>) {
      %30 = affine.max affine_map<(d0)[s0] -> (-d0 + s0, 0)>(%arg0)[%6]
      %31 = affine.max affine_map<(d0)[s0] -> (0, d0 - s0)>(%arg0)[%6]
      %32 = affine.min affine_map<(d0)[s0, s1] -> (s0 - s1, d0)>(%31)[%10, %8]
      %33 = affine.max affine_map<(d0)[s0] -> (0, d0 - s0 + 1)>(%arg0)[%6]
      %34 = affine.min affine_map<(d0)[s0, s1] -> (s0 - s1, d0)>(%33)[%10, %8]
      %35 = affine.apply affine_map<(d0, d1)[s0] -> (d0 - d1 + s0)>(%34, %32)[%30]
      %36 = arith.cmpi sle, %30, %c0 : index
      %37 = arith.cmpi sgt, %35, %c0 : index
      %38 = arith.andi %36, %37 : i1
      %39 = affine.apply affine_map<()[s0] -> (-s0)>()[%30]
      %40 = scf.for %arg2 = %c0 to %c19 step %c1 iter_args(%arg3 = %arg1) -> (tensor<1x19x19x64xf32>) {
        %41 = affine.max affine_map<(d0)[s0] -> (-d0 + s0, 0)>(%arg2)[%14]
        %42 = affine.max affine_map<(d0)[s0] -> (0, d0 - s0)>(%arg2)[%14]
        %43 = affine.min affine_map<(d0)[s0, s1] -> (s0 - s1, d0)>(%42)[%18, %16]
        %44 = affine.max affine_map<(d0)[s0] -> (0, d0 - s0 + 1)>(%arg2)[%14]
        %45 = affine.min affine_map<(d0)[s0, s1] -> (s0 - s1, d0)>(%44)[%18, %16]
        %46 = affine.apply affine_map<(d0, d1)[s0] -> (d0 - d1 + s0)>(%45, %43)[%41]
        %47 = arith.cmpi sle, %41, %c0 : index
        %48 = arith.cmpi sgt, %46, %c0 : index
        %49 = arith.andi %47, %48 : i1
        %50 = arith.andi %38, %49 : i1
        %51 = affine.apply affine_map<()[s0] -> (-s0)>()[%41]
        %52 = scf.for %arg4 = %c0 to %c64 step %c4 iter_args(%arg5 = %arg3) -> (tensor<1x19x19x64xf32>) {
          %53 = scf.if %50 -> (vector<4xf32>) {
            %55 = arith.addi %39, %32 : index
            %56 = arith.addi %51, %43 : index
            %57 = vector.transfer_read %22[%24, %55, %56, %arg4], %cst_0 {in_bounds = [true]} : tensor<?x?x?x64xf32>, vector<4xf32>
            scf.yield %57 : vector<4xf32>
          } else {
            scf.yield %cst : vector<4xf32>
          }
          %54 = vector.transfer_write %53, %25[%c0, %c0, %c0, %c0] {in_bounds = [true]} : vector<4xf32>, tensor<1x1x1x4xf32>
          %inserted_slice = tensor.insert_slice %54 into %arg5[0, %arg0, %arg2, %arg4] [1, 1, 1, 4] [1, 1, 1, 1] : tensor<1x1x1x4xf32> into tensor<1x19x19x64xf32>
          scf.yield %inserted_slice : tensor<1x19x19x64xf32>
        }
        scf.yield %52 : tensor<1x19x19x64xf32>
      }
      scf.yield %40 : tensor<1x19x19x64xf32>
    }
    %27 = affine.apply affine_map<()[s0] -> ((s0 floordiv 6) floordiv 6)>()[%workgroup_id_x]
    %28 = affine.apply affine_map<()[s0] -> ((s0 floordiv 6) * 19 - ((s0 floordiv 6) floordiv 6) * 114)>()[%workgroup_id_x]
    %29 = affine.apply affine_map<()[s0] -> (s0 * 19 - (s0 floordiv 6) * 114)>()[%workgroup_id_x]
    flow.dispatch.tensor.store %26, %1, offsets = [%27, %28, %29, 0], sizes = [1, 19, 19, 64], strides = [1, 1, 1, 1] : tensor<1x19x19x64xf32> -> !flow.dispatch.tensor<writeonly:tensor<1x114x114x64xf32>>
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:132]: I filed an issue for this: https://github.com/openxla/iree/issues/13466

**Comment by hanhanW** [comment]
It's working now.. Maybe I should move the bufferization change to another PR.. working on it.

**Comment by benvanik** [comment]
Closing stale PR.

**Pull request closed without merging**
